drivers/ethernet/phy_xlnx_gem.c

```diff
@@ -0,0 +1,975 @@
+		dev_data->phy_addr, PHY_MRVL_COPPER_STATUS_REGISTER);
+
+	return ((phy_data >> PHY_MRVL_COPPER_LINK_STATUS_BIT_SHIFT) & 0x0001);
+}
+
+/**
+ * @brief Marvell Alaska PHY link speed polling function
+ * Link speed polling function for the Marvell Alaska PHY series
+ *
+ * @param dev Pointer to the device data
+ * @return    Enum containing the current link speed reported by the PHY
+ */
+static enum eth_xlnx_link_speed phy_xlnx_gem_marvell_alaska_poll_lspd(
+	const struct device *dev)
+{
+	const struct eth_xlnx_gem_dev_cfg *dev_conf = DEV_CFG(dev);
+	struct eth_xlnx_gem_dev_data *dev_data  = DEV_DATA(dev);
+	enum eth_xlnx_link_speed link_speed = LINK_DOWN;
+	uint16_t phy_data = 0;
+
+	/*
+	 * Current link speed is obtained from:
+	 * Page 0, register address 17 = Copper Specific Status Register 1
+	 * bits [15 .. 14] = Speed.
+	 */
+	phy_data   = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
+		dev_data->phy_addr, PHY_MRVL_COPPER_STATUS_1_REGISTER);
+	phy_data >>= PHY_MRVL_LINK_SPEED_SHIFT;
+	phy_data  &= PHY_MRVL_LINK_SPEED_MASK;
+
+	/*
+	 * Link speed bit masks: comp. datasheet, table 77 @ description
+	 * of the 'Speed' bits.
+	 */
+	switch (phy_data) {
+	case PHY_MRVL_LINK_SPEED_10MBIT:
+		link_speed = LINK_10MBIT;
+		break;
+	case PHY_MRVL_LINK_SPEED_100MBIT:
+		link_speed = LINK_100MBIT;
+		break;
+	case PHY_MRVL_LINK_SPEED_1GBIT:
+		link_speed = LINK_1GBIT;
+		break;
+	default:
+		link_speed = LINK_DOWN;
+		break;
+	};
+
+	return link_speed;
+}
+
+/**
+ * @brief TI TLK105 & DP83822 PHY reset function
+ * Reset function for the TI TLK105 & DP83822 PHYs
+ *
+ * @param dev Pointer to the device data
+ */
+static void phy_xlnx_gem_ti_dp83822_reset(const struct device *dev)
+{
+	const struct eth_xlnx_gem_dev_cfg *dev_conf = DEV_CFG(dev);
+	struct eth_xlnx_gem_dev_data *dev_data = DEV_DATA(dev);
+	uint16_t phy_data = 0;
+	uint32_t retries = 0;
+
+	phy_data = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
+		dev_data->phy_addr, PHY_TI_REG_BMCR);
+	phy_data |= 0x8000; /* Reset bit */
+	phy_xlnx_gem_mdio_write(dev_conf->base_addr,
+		dev_data->phy_addr, PHY_TI_REG_BMCR, phy_data);
+
+	while (((phy_data & 0x8000) != 0) && (retries++ < 10)) {
+		phy_data = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
+			dev_data->phy_addr, PHY_TI_REG_BMCR);
+	}
+	if (retries == 10) {
+		LOG_ERR("%s reset PHY address %hhu (TI DP83822) timed out",
+			dev->name, dev_data->phy_addr);
+	}
+}
+
+/**
+ * @brief TI TLK105 & DP83822 PHY configuration function
+ * Configuration function for the TI TLK105 & DP83822 PHYs
+ *
+ * @param dev Pointer to the device data
+ */
+static void phy_xlnx_gem_ti_dp83822_cfg(const struct device *dev)
+{
+	const struct eth_xlnx_gem_dev_cfg *dev_conf = DEV_CFG(dev);
+	struct eth_xlnx_gem_dev_data *dev_data = DEV_DATA(dev);
+	uint16_t phy_data = 0x0001;
+
+	if (dev_conf->enable_fdx) {
+		if (dev_conf->max_link_speed == LINK_100MBIT) {
+			/* Advertise 100BASE-TX, full duplex */
+			phy_data |= (1 << 8);
```

Comment:
all of these bits should be macros as well.  Pretty much throughout this whole function there are open coded bits that should probably be macros.